src/Voron/Impl/Journal/JournalWriter.cs

```diff
@@ -28,49 +28,87 @@ public unsafe class JournalWriter : IJournalWriter
         private bool _workingSetQuotaLogged = false;
+        private readonly StorageEnvironmentSynchronization _writeSynchronization;
+        private readonly long _maxNumberOf4KbInSingleWrite;
 
         public int NumberOfAllocated4Kb { get; }
         public bool Disposed => _disposed.IsRaised();
         public VoronPathSetting FileName { get; }
         public bool DeleteOnClose { get; set; }
 
-        public JournalWriter(StorageEnvironmentOptions options, VoronPathSetting filename, long size, PalFlags.JournalMode mode = PalFlags.JournalMode.Safe)
+        public JournalWriter(StorageEnvironmentOptions options, VoronPathSetting filename, long size, StorageEnvironmentSynchronization writeSynchronization = null, PalFlags.JournalMode mode = PalFlags.JournalMode.Safe)
         {
-            _options = options;
+            _options = options;            
             FileName = filename;
             _log = LoggingSource.Instance.GetLogger<JournalWriter>(options.BasePath.FullPath);
 
+            if (writeSynchronization == null)
+                writeSynchronization = new StorageEnvironmentSynchronization();
+
+            // The write synchronization mechanism will require a max size and to be aligned to 4kb.  
+            _writeSynchronization = writeSynchronization;
+            _maxNumberOf4KbInSingleWrite = writeSynchronization.MaxWriteSize / 4096;
+
             var result = Pal.rvn_open_journal_for_writes(filename.FullPath, mode, size, options.SupportDurabilityFlags, out _writeHandle, out var actualSize, out var error);
             if (result != PalFlags.FailCodes.Success)
                 PalHelper.ThrowLastError(result, error, $"Attempted to open journal file - Path: {filename.FullPath} Size :{size}");
 
             NumberOfAllocated4Kb = (int)(actualSize / (4 * Constants.Size.Kilobyte));
         }
 
-        public TimeSpan Write(long posBy4Kb, byte* p, int numberOf4Kb)
+        public TimeSpan Write(long posBy4Kb, byte* buffer, int numberOf4Kb)
         {
             Debug.Assert(_options.IoMetrics != null);
 
             Stopwatch sp;
-
             using (var metrics = _options.IoMetrics.MeterIoRate(FileName.FullPath, IoMetrics.MeterType.JournalWrite, numberOf4Kb * 4L * Constants.Size.Kilobyte))
             {
                 sp = Stopwatch.StartNew();
 
-                var result = Pal.rvn_write_journal(_writeHandle, p, numberOf4Kb * 4L * Constants.Size.Kilobyte, posBy4Kb * 4L * Constants.Size.Kilobyte, out var error);
-                if (result != PalFlags.FailCodes.Success)
-                    PalHelper.ThrowLastError(result, error, $"Attempted to write to journal file - Path: {FileName.FullPath} Size: {numberOf4Kb * 4L * Constants.Size.Kilobyte}, numberOf4Kb={numberOf4Kb}");
+                // We will separate them in chunks of max size. 
+                long iterations = numberOf4Kb / _maxNumberOf4KbInSingleWrite;
+                long leftovers = numberOf4Kb % _maxNumberOf4KbInSingleWrite;
```

Comment:
You can assert that this `leftovers` is zero here, we should always be passing 4KB aligned buffers here.

Reply:
Not really, those are 4kb aligned blocks :) a zero on both means no write in effect.

Reply:
The key here is that we want to remove the code to handle the left overs

Reply:
Leftovers are 4kb aligned blocks that need to be written. Unless what you mean is that we could simplify that code instead of using 2 sequences, use a single that handle both cases.